README.md
# react-code-input
React component for entering and validating PIN code.

[![CircleCI](https://circleci.com/gh/40818419/react-code-input.svg?style=svg)](https://circleci.com/gh/40818419/react-code-input) [![npm version](https://badge.fury.io/js/react-code-input.svg)](https://badge.fury.io/js/react-code-input) [![codecov](https://codecov.io/gh/40818419/react-code-input/branch/master/graph/badge.svg)](https://codecov.io/gh/40818419/react-code-input)

##### Live example [here](https://40818419.github.io/react-code-input)

## Installation

`npm i --save react-code-input`

## Usage

### Numeric input:
![Numeric input](https://cloud.githubusercontent.com/assets/2235499/21593670/eafea336-d119-11e6-9d4b-738141f24b44.png)
```js
...
<ReactCodeInput type='number' fields={6} />
...
```

### Text input:
![Text input](https://cloud.githubusercontent.com/assets/2235499/21593708/61cebee2-d11a-11e6-9b8d-e99dbeeec23b.png)
```js
...
<ReactCodeInput type='text' fields={6} />
...
```
### Password input:
![Password input](https://cloud.githubusercontent.com/assets/2235499/21593710/65ed7252-d11a-11e6-995f-f0127af5bd9f.png)
```js
...
<ReactCodeInput type='password' fields={6} />
...
```
### Numeric input with options:
![Numeric input with options](https://cloud.githubusercontent.com/assets/2235499/21593674/f3bb887c-d119-11e6-8b3f-ba478a0f4692.png)
```js
import { reactCodeInput } from 'CodeInputField.scss'
...
const props = {
  className: reactCodeInput,
  inputStyle: {
    fontFamily: 'monospace',
    margin:  '4px',
    MozAppearance: 'textfield',
    width: '15px',
    borderRadius: '3px',
    fontSize: '14px',
    height: '26px',
    paddingLeft: '7px',
    backgroundColor: 'black',
    color: 'lightskyblue',
    border: '1px solid lightskyblue'
  },
  inputStyleInvalid: {
    fontFamily: 'monospace',
    margin:  '4px',
    MozAppearance: 'textfield',
    width: '15px',
    borderRadius: '3px',
    fontSize: '14px',
    height: '26px',
    paddingLeft: '7px',
    backgroundColor: 'black',
    color: 'red',
    border: '1px solid red'
  }
}
...
<ReactCodeInput type='number' fields={6} {...props}/>
...
```
### Usage with next.js
```js
import dynamic from 'next/dynamic';

const ReactCodeInput = dynamic(import('react-code-input'));
...
<ReactCodeInput type='number' fields={6} {...props}/>
...
```

## Props:

| Property | Type | Description |
|:---|:---|:---|
| type | string | Only types like: `text`, `number`, `password` and `tel` are accepted.|
| fields | number | Allowed amount of characters to enter. |
| value | string | Setting the value of code input field. |
| name | string | Setting the name of component. |
| onChange | func | Function, which is called whenever there is a change of value in the input box. |
| touch | func | Marks the given fields as "touched" to show errors. |
| untouch | func | Clears the "touched" flag for the given fields. |
| className | string | Add classname to the root element. |
| style | object | Setting the styles of container element. |
| inputStyle | object | Setting the styles of each input field. |
| inputStyleInvalid | object | Setting the styles of each input field if `isValid` prop is `false`. |
| isValid | bool | Returns true if an input element contains valid data. |
| disabled | bool | When present, it specifies that the element should be disabled. |
| autoFocus | bool | Setup autofocus on the first input, `true` by default. |
| filterKeyCodes | array | Filter characters on key down. |
| filterChars | array | Filter characters. | default: ['-', '.'] |
| pattern | string | The pattern prop specifies a regular expression that the <input> element's value is checked against. |
| inputMode | string | The inputMode prop tells the browser on devices with dynamic keyboards which keyboard to display. |
| required | bool | The HTML/React `required` prop for input elements, determines if input is required before form submit, `false` by default. |

## Compatible with
[`redux-form`](https://github.com/erikras/redux-form) from [erikras](https://github.com/erikras)

[`next.js`](https://github.com/zeit/next.js) from [zeit](https://github.com/zeit)

## License
MIT


CHANGELOG.md
#### 3.9.3 (2019-11-25)

##### New Features

*  add required prop to input elements ([0056e794](https://github.com/40818419/react-code-input/commit/0056e794cd7878ce6b6922b83f74211443d684af))

#### 3.9.2 (2019-11-20)

#### 3.8.1 (2018-11-02)

##### Documentation Changes

* **next-js:**  added usage instructions with next.js ([e8b3daad](https://github.com/40818419/react-code-input/commit/e8b3daad4d7002f75bb5554f7e19fcbe135a1d24))

### 3.8.0 (2018-09-27)

### 3.7.0 (2018-09-24)

##### Chores

* **vulns:**  and fix them ([#49](https://github.com/40818419/react-code-input/pull/49)) ([0add2276](https://github.com/40818419/react-code-input/commit/0add2276fb63f2f789fa944535238aa0837ff308))
* **audit-vulnerabilities:**  and fix them ([#48](https://github.com/40818419/react-code-input/pull/48)) ([454e084b](https://github.com/40818419/react-code-input/commit/454e084b116d57bc35f74fc862e4bd144aaf5f7c))

##### Documentation Changes

* **forceUppercase:**  added onChange event ([7364794a](https://github.com/40818419/react-code-input/commit/7364794ac040d77f850b17a7ccbcbb4d2beaa316))
* **forceUpdate:**
  *  add story ([09c32c86](https://github.com/40818419/react-code-input/commit/09c32c86be1f100f7a195a0344cdaa32f3e3a3c9))
  *  add story ([bf726741](https://github.com/40818419/react-code-input/commit/bf726741a27f8d17182fba451ec277a84b7fd8ce))
* **build:**  added new storybook build ([99dc64fb](https://github.com/40818419/react-code-input/commit/99dc64fb9390a25541b8b1c21f601a7644312b37))

##### Bug Fixes

* **onchange:**  trigger onChange on ampty value ([#54](https://github.com/40818419/react-code-input/pull/54)) ([524717bf](https://github.com/40818419/react-code-input/commit/524717bfe04ea1e7ba1295d3431b809647981058))
* **docs:**  fix build ([64bfd5c7](https://github.com/40818419/react-code-input/commit/64bfd5c7264d33329a34b8634216b017164e46f1))

##### Other Changes

* **storybook:**  storybook added ([#50](https://github.com/40818419/react-code-input/pull/50)) ([5fdb6e81](https://github.com/40818419/react-code-input/commit/5fdb6e813f22cbbca2b96ea1d5c9bacaaa73a74d))

### 3.6.0 (2018-09-06)

### 3.5.0 (2018-07-02)

##### New Features

* **pattern-input-mode:**  added pattern and inputMode props ([#39](https://github.com/40818419/react-code-input/pull/39)) ([0c293463](https://github.com/40818419/react-code-input/commit/0c2934639278b83eb05dd362427bb23bf483fb66))

##### Bug Fixes

* **ssr:**  Fix server side rendering ([#37](https://github.com/40818419/react-code-input/pull/37)) ([630bed85](https://github.com/40818419/react-code-input/commit/630bed85be597bc74707ae2871e0a54f37f57a92))

### 3.3.0 (2018-06-20)

##### New Features

* **prevent-chars:**  Filter characters on key down ([#35](https://github.com/40818419/react-code-input/pull/35)) ([b6a00d90](https://github.com/40818419/react-code-input/commit/b6a00d90fdf803d319adadd0938d7fe6d53479d7))

#### 3.2.1 (2018-06-11)

##### Continuous Integration

* **add-release-to-pipeline:**  release notes ([#33](https://github.com/40818419/react-code-input/pull/33)) ([0dbb588a](https://github.com/40818419/react-code-input/commit/0dbb588a78d3abdab4a34afb9dd46f5fb570ddb8))

##### Documentation Changes

*  fix typo ([#16](https://github.com/40818419/react-code-input/pull/16)) ([a4625843](https://github.com/40818419/react-code-input/commit/a4625843027d851252690eedd7acef2730ffe321))



src/ReactCodeInput.d.ts
// Type definitions for react-code-input 3.8.2
// Project: React Code Input
// Definitions by: Siraj Alam https://github.com/sirajalam049

import React, { Component } from "react";

export type InputModeTypes =
  | "verbatim"
  | "latin"
  | "latin-name"
  | "latin-prose"
  | "full-width-latin"
  | "kana"
  | "kana-name"
  | "katakana"
  | "numeric"
  | "tel"
  | "email"
  | "url";

export interface ReactCodeInputProps {
  // Type of input accept
  type?: "text" | "number" | "password" | "tel";

  // Allowed amount of characters to enter.
  fields?: number;

  // Value of the input
  value?: string;

  // Get the full value of the input on every change
  onChange?: (value: string) => void;

  // Setting the name of component.
  name: string;

  // Marks the given fields as "touched" to show errors.
  touch?: (name: string) => void;

  // Clears the "touched" flag for the given fields.
  untouch?: (name: string) => void;

  // Add classname to the root element.
  className?: string;

  isValid?: boolean;

  // When present, it specifies that the element should be disabled.
  disabled?: boolean;

  // Setting the styles of container element.
  style?: React.CSSProperties;

  // Setting the styles of each input field.
  inputStyle?: React.CSSProperties;

  // Setting the styles of each input field if isValid prop is false.
  inputStyleInvalid?: React.CSSProperties;

  // Setup autofocus on the first input, true by default.
  autoFocus?: boolean;

  //
  forceUppercase?: boolean;

  // Filter characters on key down.
  filterKeyCodes?: Array<number>;

  // Filter characters.
  filterChars?: Array<string>;

  // Filter above acts as blacklist if true, whitelist if false, true by default.
  filterCharsIsBlacklist?: boolean;

  // The pattern prop specifies a regular expression that the element's value is checked against.
  pattern?: string;

  // The inputMode prop tells the browser on devices with dynamic keyboards which keyboard to display.
  inputMode: InputModeTypes;

  // The HTML/React `required` prop for input elements, determines if input is required before form submit, `false` by default.
  required?: boolean;
}

declare class ReactCodeInput extends Component<ReactCodeInputProps, any> {
  constructor(props: ReactCodeInputProps);
}

export default ReactCodeInput;


src/ReactCodeInput.js
/**
 * Copyright (c) 2016 Konstantin Kulinicenko.
 * Licensed under the MIT License (MIT), see
 * https://github.com/40818419/react-code-input
 */

import React, { Component } from "react";
import PropTypes from "prop-types";
import classNames from "classnames";
import { uuidv4 } from "./utils";

const BACKSPACE_KEY = 8;
const LEFT_ARROW_KEY = 37;
const UP_ARROW_KEY = 38;
const RIGHT_ARROW_KEY = 39;
const DOWN_ARROW_KEY = 40;
const E_KEY = 69;

class ReactCodeInput extends Component {
  constructor(props) {
    super(props);

    const {
      fields,
      type,
      isValid,
      disabled,
      filterKeyCodes,
      forceUppercase
    } = props;
    let { value } = props;

    if (forceUppercase) {
      value = value.toUpperCase();
    }

    this.state = {
      value,
      fields,
      type,
      input: [],
      isValid,
      disabled,
      filterKeyCodes,
      defaultInputStyle: {
        fontFamily: "monospace",
        MozAppearance: "textfield",
        borderRadius: "6px",
        border: "1px solid",
        boxShadow: "0px 0px 10px 0px rgba(0,0,0,.10)",
        margin: "4px",
        paddingLeft: "8px",
        paddingRight: 0,
        width: "36px",
        height: "42px",
        fontSize: "32px",
        boxSizing: "border-box"
      }
    };

    for (let i = 0; i < Number(this.state.fields); i += 1) {
      if (i < 32) {
        const value = this.state.value[i] || "";
        this.state.input.push(value);
      }
    }

    this.textInput = [];

    this.uuid = uuidv4();
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    this.setState({
      isValid: nextProps.isValid,
      value: nextProps.value,
      disabled: nextProps.disabled
    });
  }

  handleBlur(e) {
    this.handleTouch(e.target.value);
  }

  handleTouch(value) {
    const { touch, untouch, name } = this.props;

    if (typeof touch === "function" && typeof untouch === "function") {
      if (value === "") {
        touch(name);
      } else {
        untouch(name);
      }
    }
  }

  handleChange(e) {
    const { filterChars, filterCharsIsBlacklist } = this.props;

    let value = String(e.target.value);

    if (this.props.forceUppercase) {
      value = value.toUpperCase();
    }

    if (this.state.type === "number") {
      value = value.replace(/[^\d]/g, "");
    }

    /** Filter Chars */
    value = value
      .split("")
      .filter(currChar =>
        filterCharsIsBlacklist
          ? !filterChars.includes(currChar)
          : filterChars.includes(currChar)
      )
      .join("");

    let fullValue = value;

    if (value !== "") {
      const input = this.state.input.slice();

      if (value.length > 1) {
        value.split("").map((chart, i) => {
          if (Number(e.target.dataset.id) + i < this.props.fields) {
            input[Number(e.target.dataset.id) + i] = chart;
          }
          return false;
        });
      } else {
        input[Number(e.target.dataset.id)] = value;
      }

      input.map((s, i) => {
        if (this.textInput[i]) {
          this.textInput[i].value = s;
        }
        return false;
      });

      const newTarget = this.textInput[
        e.target.dataset.id < input.length
          ? Number(e.target.dataset.id) + 1
          : e.target.dataset.id
      ];

      if (newTarget) {
        newTarget.focus();
        newTarget.select();
      }

      fullValue = input.join("");

      this.setState({ value: input.join(""), input });
    }

    if (this.props.onChange && fullValue) {
      this.props.onChange(fullValue);
    }

    this.handleTouch(fullValue);
  }

  handleKeyDown(e) {
    const target = Number(e.target.dataset.id),
      nextTarget = this.textInput[target + 1],
      prevTarget = this.textInput[target - 1];

    let input, value;

    if (this.state.filterKeyCodes.length > 0) {
      this.state.filterKeyCodes.map(item => {
        if (item === e.keyCode) {
          e.preventDefault();
          return true;
        }
      });
    }

    switch (e.keyCode) {
      case BACKSPACE_KEY:
        e.preventDefault();
        this.textInput[target].value = "";
        input = this.state.input.slice();
        input[target] = "";
        value = input.join("");

        this.setState({ value, input });
        if (this.textInput[target].value === "") {
          if (prevTarget) {
            prevTarget.focus();
            prevTarget.select();
          }
        }
        if (this.props.onChange) {
          this.props.onChange(value);
        }
        break;

      case LEFT_ARROW_KEY:
        e.preventDefault();
        if (prevTarget) {
          prevTarget.focus();
          prevTarget.select();
        }
        break;

      case RIGHT_ARROW_KEY:
        e.preventDefault();
        if (nextTarget) {
          nextTarget.focus();
          nextTarget.select();
        }
        break;

      case UP_ARROW_KEY:
        e.preventDefault();
        break;

      case DOWN_ARROW_KEY:
        e.preventDefault();
        break;

      case E_KEY: // This case needs to be handled because of https://stackoverflow.com/questions/31706611/why-does-the-html-input-with-type-number-allow-the-letter-e-to-be-entered-in
        if (e.target.type === "number") {
          e.preventDefault();
          break;
        }
        break;

      default:
        break;
    }

    this.handleTouch(value);
  }

  render() {
    const {
        className,
        style = {},
        inputStyle = {},
        inputStyleInvalid = {},
        type,
        autoFocus,
        pattern,
        inputMode,
        required
      } = this.props,
      { disabled, input, isValid, defaultInputStyle } = this.state,
      styles = {
        container: { display: "inline-block", ...style },
        input: isValid ? inputStyle : inputStyleInvalid
      };

    if (!className && Object.keys(inputStyle).length === 0) {
      Object.assign(inputStyle, {
        ...defaultInputStyle,
        color: "black",
        backgroundColor: "white",
        borderColor: "lightgrey"
      });
    }

    if (!className && Object.keys(inputStyleInvalid).length === 0) {
      Object.assign(inputStyleInvalid, {
        ...defaultInputStyle,
        color: "#b94a48",
        backgroundColor: "#f2dede",
        borderColor: "#eed3d7"
      });
    }

    if (disabled) {
      Object.assign(styles.input, {
        cursor: "not-allowed",
        color: "lightgrey",
        borderColor: "lightgrey",
        backgroundColor: "#efeff1"
      });
    }

    return (
      <div
        className={classNames(className, "react-code-input")}
        style={styles.container}
      >
        {input.map((value, i) => {
          return (
            <input
              ref={ref => {
                this.textInput[i] = ref;
              }}
              id={`${this.uuid}-${i}`}
              data-id={i}
              autoFocus={autoFocus && i === 0 ? "autoFocus" : ""}
              value={value}
              key={`input_${i}`}
              type={type}
              min={0}
              max={9}
              maxLength={input.length === i + 1 ? 1 : input.length}
              style={styles.input}
              autoComplete="off"
              onFocus={e => e.target.select(e)}
              onBlur={e => this.handleBlur(e)}
              onChange={e => this.handleChange(e)}
              onKeyDown={e => this.handleKeyDown(e)}
              disabled={disabled}
              data-valid={isValid}
              pattern={pattern}
              inputMode={inputMode}
              required={required}
            />
          );
        })}
      </div>
    );
  }
}

ReactCodeInput.defaultProps = {
  autoFocus: true,
  isValid: true,
  disabled: false,
  forceUppercase: false,
  fields: 4,
  value: "",
  type: "text",
  filterKeyCodes: [189, 190],
  filterChars: ["-", "."],
  filterCharsIsBlacklist: true,
  required: false,
};

ReactCodeInput.propTypes = {
  type: PropTypes.oneOf(["text", "number", "password", "tel"]),
  fields: PropTypes.number,
  value: PropTypes.string,
  onChange: PropTypes.func,
  name: PropTypes.string,
  touch: PropTypes.func,
  untouch: PropTypes.func,
  className: PropTypes.string,
  isValid: PropTypes.bool,
  disabled: PropTypes.bool,
  style: PropTypes.object,
  inputStyle: PropTypes.object,
  inputStyleInvalid: PropTypes.object,
  autoFocus: PropTypes.bool,
  forceUppercase: PropTypes.bool,
  filterKeyCodes: PropTypes.array,
  filterChars: PropTypes.array,
  filterCharsIsBlacklist: PropTypes.bool,
  pattern: PropTypes.string,
  inputMode: PropTypes.oneOf([
    "verbatim",
    "latin",
    "latin-name",
    "latin-prose",
    "full-width-latin",
    "kana",
    "kana-name",
    "katakana",
    "numeric",
    "tel",
    "email",
    "url"
  ]),
  required: PropTypes.bool,
};

export default ReactCodeInput;
